crates/sui-faucet/src/server.rs

```diff
@@ -35,12 +35,203 @@ use tracing::{info, warn};
+
+static DISCORD_BOT: Lazy<Option<String>> = Lazy::new(|| std::env::var("DISCORD_BOT").ok());
+
+static TURNSTILE_SECRET_KEY: Lazy<Option<String>> =
+    Lazy::new(|| std::env::var("TURNSTILE_SECRET_KEY").ok());
+
+type IPAddr = String;
+
+/// Keep track of every IP address' requests.
+#[derive(Debug)]
+struct RequestsManager {
+    pub data: Mutex<BTreeMap<IPAddr, RequestInfo>>,
+    reset_time_interval_secs: u64,
+    max_requests_per_ip: u32,
+}
+
+/// Request's metadata
+#[derive(Debug, Clone)]
+struct RequestInfo {
+    expires_at: u64,
+    requests_used: u32,
+    total_requests_allowed: u32,
+}
+
+/// Struct to deserialize token verification response from Cloudflare
+#[derive(Deserialize, Debug)]
+struct TokenVerification {
+    success: bool,
+    #[serde(rename = "error-codes")]
+    error_codes: Vec<String>,
+}
+
+impl RequestsManager {
+    /// Initialize a new RequestsManager the default values.
+    fn new() -> Self {
+        Self {
+            data: Mutex::new(BTreeMap::new()),
+            reset_time_interval_secs: RESET_TIME_INTERVAL_SECS,
+            max_requests_per_ip: MAX_REQUESTS_PER_IP,
+        }
+    }
+
+    #[cfg(test)]
+    fn new_with_limits(max_requests_per_ip: u32, reset_time_interval_secs: u64) -> Self {
+        Self {
+            data: Mutex::new(BTreeMap::new()),
+            reset_time_interval_secs,
+            max_requests_per_ip,
+        }
+    }
+
+    /// Get the current timestamp in seconds
+    fn current_timestamp_secs() -> u64 {
+        SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .expect("Time went backwards")
+            .as_secs()
+    }
+
+    /// Checks if the user-agent is present, and if it is the expected discord bot user-agent. This
+    /// includes a unique password that the bot has set in the user-agent header.
+    fn is_discord_bot(&self, user_agent: Option<&HeaderValue>) -> Result<bool, FaucetError> {
+        if let (Some(discord_bot), Some(v)) = (DISCORD_BOT.as_ref(), user_agent) {
+            let header = v
+                .to_str()
+                .map_err(|e| FaucetError::InvalidUserAgent(e.to_string()))?;
+            Ok(header == format!("discord-bot-{}", discord_bot))
+        } else {
+            Ok(false)
+        }
+    }
+
+    /// Validates a token
+    /// - against Cloudflare turnstile's server to ensure token was issued by turnstile
+    /// - against the IP address' request count
+    async fn validate_token(
+        &self,
+        url: &str,
+        addr: SocketAddr,
+        token: &str,
+    ) -> Result<(), (StatusCode, FaucetError)> {
+        let turnstile_key = TURNSTILE_SECRET_KEY.as_ref().unwrap().as_str();
+        let req = reqwest::Client::new();
+        let params = [
+            ("secret", turnstile_key),
+            ("response", token),
+            ("remoteip", &addr.ip().to_string()),
+        ];
+
+        // Make the POST request
+        let response = req.post(url).form(&params).send().await;
```

Comment:
nit: The code from this point is drifting to the left and the error cases for outer nesting levels are quite far removed from the test they correspond to. Can you shift things into an early return style?

```rust
let resp = match req.post(url).from(&params).send().await {
    Ok(resp) => resp,
    Err(e) => ...
};

if !resp.status.is_success() {
    return Err(...);
}

let body = match resp.json::<TokenVerification>().await {
    Ok(body) => body,
    Err(e) => ...
};

// etc ...
```